### PR TITLE
write operative gin confs to disk

### PR DESCRIPTION
**Pull request by hnyu**
Write a gin configration to a file instead of just copying. Because the user can
    1) manually change the gin confs after loading a conf file into the code, or
    2) include a gin file in another gin file while only the latter might be
       copied to `root_dir`.
So here we just dump the actual used gin conf string to a file. 

We should always write both operative and inoperative confs because for some confs that are indeed used by subprocesses, they are always shown as "inoperative".

### CODE REVIEW

File: alf/utils/common.py [hunk 1]
```diff
@@ -721,3 +721,67 @@ def get_action_spec():
     """
     assert _env, "set a global env by `set_global_env` before using the function"
     return _env.action_spec()
+
+
+def extract_spec(nest):
```
Comment (emailweixu) [inline alf/utils/common.py]: extract_spec in off_policy_driver should be changed to use this one.

File: alf/utils/common.py [hunk 2]
```diff
@@ -721,3 +721,67 @@ def get_action_spec():
...
+    """
+    Extract tensor spec for each element of a nested structure.
+    It assumes that the first dimension of each element is the batch size.
```
Comment (emailweixu) [inline alf/utils/common.py]: Arguments: and Returns:

File: alf/bin/train.py [hunk 3]
```diff
@@ -76,8 +76,6 @@ def train_eval(root_dir):
 
 def main(_):
     gin_file = common.get_gin_file()
-    if FLAGS.gin_file:
-        common.copy_gin_configs(FLAGS.root_dir, gin_file)
```
Comment (emailweixu) [inline alf/bin/train.py]: This change will make continuing training or playing without gin_file fail.

Comment (witwolf) [reply alf/bin/train.py]: should copy gin files to root_dir , `operative.gin`  do not include all operative config actually (sub processes #193 )

Comment (hnyu) [reply alf/bin/train.py]: > should copy gin files to root_dir , `operative.gin` do not include all operative config actually (sub processes #193 )

I've updated the PR description for the motivation. For the subprocess issue, I think we should create a dummy environment inside the main process.

Comment (hnyu) [reply alf/bin/train.py]: > This change will make continuing training or playing without gin_file fail.

Why? If FLAGS.gin_file is None, then nothing happens. `gin_file` will be globbed inside `root_dir`

Comment (emailweixu) [reply alf/bin/train.py]: You need to copy inoperative configs

File: alf/utils/adaptive_normalizer.py [hunk 4]
```diff
@@ -49,19 +49,24 @@ def __init__(self,
         super(AdaptiveNormalizer, self).__init__(
             tensor_spec, norm_update_rate=self._update_ema_rate)
 
-    def normalize(self, tensors):
+    def normalize(self, tensors, clip_value=1.0):
```
Comment (emailweixu) [inline alf/utils/adaptive_normalizer.py]: This averager is used in different places. I think default should be no clip.

Comment (hnyu) [reply alf/utils/adaptive_normalizer.py]: updated. But note that tf_agents does have a default clip value of 5.0 originally. Not it's set to -1.0 (ignored)

**Pull request merged**